### PR TITLE
gh-95423: Updated winreg.DeleteKeyEx documentation for 32-bit OS.

### PR DESCRIPTION
**Pull request by derekdkim**
Documentation updates to address:

Issue #95423 

Please let me know if my wording could use some improvement or if I missed any other part of the documentation.

### CODE REVIEW

**Comment by zooba** [comment]
Thanks, it's a good start.

We don't need the note about `NT 5.2` in here, because that applies to all of our supported platforms. We can also really simplify the note about 32-bit OS, since that's a vanishingly small number of machines these days.

The aim of our docs is to help Python developers use Python correctly. We don't need to provide historical or internal details of the platform, except where they are directly relevant to decisions that Python devs will have to make (and we are confident they won't change on us). Hopefully that helps guide your revisions.

**Comment by derekdkim** [comment]
Thanks for the feedback @zooba. I tried my best to cut down some of the bloat. Let me know if you think it can be simplified further.

**Comment by zooba** [comment]
Ah yeah, because you modified the docstrings, you need to run `python Tools\clinic\clinic.py PC\winreg.c` to update the generated code that includes them. Once you've done that, it should pass that check.

**Comment by zooba** [comment]
For future, please don't force push once people have started reviewing. We always squash merge and edit commit messages, so no need to worry about that, and force pushes just break incremental reviews (i.e. I can't see what changes you just made, I have to start reading again).

File: Doc/library/winreg.rst [hunk 1]
```diff
@@ -159,8 +161,12 @@ This module offers the following functions:
...
 
-   *access* is an integer that specifies an access mask that describes the desired
-   security access for the key.  Default is :const:`KEY_WOW64_64KEY`.  See
+   *access* is an integer that specifies an access mask that describes the
+   desired security access for the key.  Default is :const:`KEY_WOW64_64KEY`.
+   On 32-bit Windows, the access constants :const:`KEY_WOW64_64KEY` and
+   :const:`KEY_WOW64_32KEY` are ignored, thus 0 should be passed in
+   for this parameter as a workaround.
+   See
    :ref:`Access Rights <access-rights>` for other allowed values.
```
Comment (zooba) [inline Doc/library/winreg.rst]:
```suggestion
   On 32-bit Windows, the WOW64 constants are ignored.
   See :ref:`Access Rights <access-rights>` for other allowed values.
```

I think we can simplify down to this. There's no need to override the default value, as the flags are ignored.

File: Doc/library/winreg.rst [hunk 2]
```diff
@@ -146,7 +146,9 @@ This module offers the following functions:
 
    .. note::
       The :func:`DeleteKeyEx` function is implemented with the RegDeleteKeyEx
-      Windows API function, which is specific to 64-bit versions of Windows.
+      Windows API function, which is intended to be used for 64-bit Windows.
+      On 32-bit systems, it is strongly recommended to use :func:`DeleteKey()`
+      instead.
```
Comment (zooba) [inline Doc/library/winreg.rst]: I did some more research, and it seems that there's no distinction on any modern Windows OS. `RegDeleteKey` just calls `RegDeleteKeyEx` directly and passes in zero for the extra parameter, so there isn't really any need for this note at all.

File: Doc/library/winreg.rst [hunk 3]
```diff
@@ -658,13 +664,16 @@ For more information, see `Accessing an Alternate Registry View
 .. data:: KEY_WOW64_64KEY
 
    Indicates that an application on 64-bit Windows should operate on
-   the 64-bit registry view.
+   the 64-bit registry view. On 32-bit Windows, this constant is ignored.
+   As such, to call any function that uses this constant as a default argument
+   on a 32-bit Windows build, 0 should be passed in.
```
Comment (zooba) [inline Doc/library/winreg.rst]:
```suggestion
```

As above, no need to override the default.

File: Doc/library/winreg.rst [hunk 4]
```diff
@@ -658,13 +664,16 @@ For more information, see `Accessing an Alternate Registry View
...
 
 .. data:: KEY_WOW64_32KEY
 
    Indicates that an application on 64-bit Windows should operate on
-   the 32-bit registry view.
-
+   the 32-bit registry view. On 32-bit Windows, this constant is ignored.
+   As such, to call any function that uses this constant as a default argument
+   on a 32-bit Windows build, 0 should be passed in.
```
Comment (zooba) [inline Doc/library/winreg.rst]:
```suggestion
```

As above

File: PC/winreg.c [hunk 5]
```diff
@@ -1000,7 +1000,11 @@ winreg.DeleteKeyEx
     reserved: int = 0
         A reserved integer, and must be zero.  Default is zero.
 
-Deletes the specified key (64-bit OS only).
+Deletes the specified key (intended for 64-bit OS).
+
+While this function is intended to be used for 64-bit OS, it is also
+ available on 32-bit systems.
+To use with 32-bit Windows, 0 must be passed in as the access argument.
```
Comment (zooba) [inline PC/winreg.c]:
```suggestion
Deletes the specified key.
```

As above

File: PC/winreg.c [hunk 6]
```diff
@@ -1025,8 +1029,13 @@ winreg_DeleteKeyEx_impl(PyObject *module, HKEY key,
...
     }
-    /* Only available on 64bit platforms, so we must load it
-       dynamically. */
+    /* RegDeleteKeyExW is also available on 32-bit systems.
+     * On such systems, RegDeleteKeyExW() is functionally equivalent to
+     * RegDeleteKeyW().
+     *
+     * Note: Though it currently isn't necessary, this can be simplified to use
+     * load-time dynamic linking in the future.
+     * */
```
Comment (zooba) [inline PC/winreg.c]:
```suggestion
    /* XXX: remove dynamic load, as RegDeleteKeyExW is always available */
```

**Comment by derekdkim** [comment]
Ah sorry, my bad. Noted.

File: PC/winreg.c (hunk 6 above)
Comment (zooba) [reply PC/winreg.c]: If you'd like to also remove the dynamic load and make it just call `RegDeleteKeyExW` directly, that would be fine too. Then this comment can go away entirely.

**Review by zooba — CHANGES_REQUESTED** [review]
I checked out the current state of the Reg function, and think we can drastically simplify. There aren't really any edge cases here anymore - it's just RegDeleteKeyExW all the time, and the WOW64 constants are ignored when they aren't relevant, so you can pass in anything you like.

File: Doc/library/winreg.rst (hunk 2 above)
Comment (derekdkim) [reply Doc/library/winreg.rst]: Got it, I removed the note completely.

**Comment by derekdkim** [comment]
Thanks for the suggestions. I have made the requested changes; please review again. I think I was way too wordy and tunnel visioned on keeping the notes mentioned in the issue.

I also simplified `winreg_DeleteKeyEx_impl()` as per your suggestions. I removed the dynamic loading from the `.dll` and just called it directly similar to how `winreg_DeleteKey_impl()` is implemented. Let me know if I missed anything.

File: PC/winreg.c [hunk 7]
```diff
@@ -1013,34 +1016,15 @@ static PyObject *
...
-    if (!pfn) {
-        PyErr_SetString(PyExc_NotImplementedError,
-                                        "not implemented on this platform");
-        return NULL;
-    }
-    Py_BEGIN_ALLOW_THREADS
-    rc = (*pfn)(key, sub_key, access, reserved);
-    Py_END_ALLOW_THREADS
-
+    rc = RegDeleteKeyExW(key, sub_key, access, reserved);
```
Comment (zooba) [inline PC/winreg.c:1030]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    rc = RegDeleteKeyExW(key, sub_key, access, reserved);
    Py_END_ALLOW_THREADS
```

We still want to release and reacquire the GIL around this call, which means bringing back the BEGIN/END_ALLOW_THREADS calls.

**Review by zooba — COMMENTED** [review]
With the one change (which I'll commit myself), this looks great!

**Comment by derekdkim** [comment]
Ah thanks for the note about the GIL around `RegDeleteKeyExW()`. Is there a reason why the same doesn't apply for `RegDeleteKeyW()` in [`winreg_DeleteKey_impl`](https://github.com/python/cpython/blob/03e3cb1aa5338d8a1d4ea2019491e3f6bd350f59/PC/winreg.c#L982)

**Comment by zooba** [comment]
Hmm... that seems like an oversight. Would you like to get it? I'm fine to not worry about it and just merge as it is.

**Comment by derekdkim** [comment]
No problem, I got it.

**Comment by zooba** [comment]
Awesome work, thanks!
